Partialgwmodel

LaurelineJ opened the pull request:
Better dealing of groundwater data, especially for state groundwater model.

LaurelineJ commented:
This allows us to save partial gw model so that when we want to focus on a given state, we do not need to read the contus data. 

This is in conflict with the state level model handling of data.

jrising commented:
Hi Laureline, can you please merge in the bystate branch into this branch and update the code?

LaurelineJ commented:
I'll work on it next week.

jrising commented on src/groundwaterdata.jl:
```diff
@@ -0,0 +1,47 @@
+            # Save groundwater model
+            writetable(datapath("cache/gwmodel/dfgw$suffix.csv"), dfgw)
+            writetable(datapath("cache/gwmodel/lateralconductivity$suffix.csv"), convert(DataFrame, lateralconductivity))
+            writetable(datapath("cache/gwmodel/aquiferconnexion$suffix.csv"), convert(DataFrame, aquiferconnexion))
```
Is there a significant benefit from caching these?  It seems like either way we're just reading in a data file.

LaurelineJ replied:
It is only for when the model is run at a county-scale for a specific state that a partial gwmodel would be created ... it makes the simulation of groundwater faster ... I can delete it if you find it superfluous

LaurelineJ commented:
The main initial goal of this branch was to save a state specific groundwater model for state-specific analysis. The idea was to make the process faster and not load all the groundwater data. I was also to shorten the Groundwater.jl file to move all data reading lines to a groundwaterdata.jl script. All vector type data have been aggregated in one csv file, while the connexion and lateral conductivity matrix are kept in separate csv file. In the future, this should be improved by having only one sparse matrix. 

I have moved the data in the correct place (I hope), it is now in data/counties/gwmodel.
I have also added something to warn that the state-level groundwater model does not exist yet. Though I think I can construct something quickly by making bold assumptions on storage coefficient. Updates on this to come in the near future.

There are a few additional changes in there that are not related to the branch per say. I am not sure if they are useful or not. For instance:
- Caching: here cached_fallback(), cached_store() and the test_caching script all follow the cachepath.
- I have temporally removed the variation of groundwater cost in space until we have depth to groundwater information (rather than thickness of groundwater layer)
- simulate.jl does not necessarily read a config file if one is already loaded

jrising commented on src/optimize-waterallocation.jl:
```diff
@@ -43,7 +43,10 @@ serialize(open(datapath("extraction/waterfromgw$suffix.jld"), "w"), reshape(sol.
+    if storedcaptures != false
+        rm(datapath("extraction/captures$suffix.jld"))
+    end
 end
```
Can you explain this change?  It seems like it's saying to delete the file if it can be read, which is what the original did.

LaurelineJ replied:
I believe that if the file is non existent, it will try to delete a file that does not exist and return an error ...

jrising replied:
Okay, I see that.  But you don't want to take all the time to read the file, if you're just going to delete it.  The way to check if a file exists in julia is `isfile(path)`.

jrising commented on src/lib/datastore.jl:
```diff
@@ -51,9 +51,9 @@ function cached_fallback(filename, generate)
+        if isfile(cachepath("$filename$suffix-$confighash.jld"))
             return deserialize(open(cachepath("$filename$suffix-$confighash.jld")))
-        elseif isfile(datapath("$filename$suffix.jld"))
+        elseif isfile(cachepath("$filename$suffix.jld"))
```
You're right, this is how it should be.

LaurelineJ replied:
Of course, I got carried away with the use of cached_fallback. We are only checking if the file exists now.

jrising commented:
Seems great.  Want to merge it in?
